### PR TITLE
Runtime: Add `WatchStreamExt::subscribe`

### PR DESCRIPTION
**Pull request by danrspencer**
## Motivation

It may be desirable to have subscribe to the events from a watcher without taking ownership of the stream itself. For example this allows a controller to be created via the `applier` without consuming the original watcher stream.

## Solution

Added an extension method in `watch_ext` which produces a `stream_subscriber`. Success events (`T` must implement clone, so we can't send errors to subscribers)  are propagated to all subscribers as the original stream is consumed.

### CODE REVIEW

**Comment by nightkr** [comment]
Sadly, I'm not sure this is a good fit for us, for a couple of reasons:

- This doesn't inherently have anything to do with Kubernetes, it could be a generic stream adapter
- This implementation involves a few footguns:
  - The secondary watchers will stall if the primary watcher is never polled
  - https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html#lagging means that watchers that can't keep up with the primary consumer will miss change events

As for the clone issue, I think I'd generally rather `Arc`-wrap the messages than drop errors (or demand that users do so if required). That should also help efficiency in the fast case since K8s objects tend to be relatively costly to clone (lots of small strings and vecs that need to be allocated separately).

**Comment by danrspencer** [comment]
@nightkr see https://github.com/kube-rs/kube/pull/1128#discussion_r1093371009 for some context on this PR

**Comment by danrspencer** [comment]
To address the other comments more specifically:

- Wrapping the stream items in an Arc would change the signature of the original stream. I'm not sure if that's problematic but it would stop anything else that expects to be able to take an `Stream<Item = Result<watcher::Event<K>, watcher::Error>>` from using the new stream. Wrapping the item before pushing it into the channel to prevent >1 clone of the Kubernetes item seems reasonable though. 
- A `Store` will remain always empty if its stream is never polled, I'm unsure how the subscription not producing items is more of a foot gun than that?
- The lagging concern seems reasonable; I'm not sure of the best way to solve it. Calling `stream.subscribable()` could return a wrapper over the original stream and a single subscription stream in a tuple since the primary uses cases I see would be for a very small number of subscribers. The downside of that approach is each additional stream is another layer on top of the original stream.

**Comment by nightkr** [comment]
I'm sorry for the tone in my original reply. I don't want to shut this down, it's just a complicated topic that I'm scared of getting wrong.

I'll go over the other PR to try to understand the use case better.

> Wrapping the stream items in an Arc would change the signature of the original stream. I'm not sure if that's problematic but it would stop anything else that expects to be able to take an Stream<Item = Result<watcher::Event<K>, watcher::Error>> from using the new stream. Wrapping the item before pushing it into the channel to prevent >1 clone of the Kubernetes item seems reasonable though.

Yeah, that's a concern. In theory most consumers "should" be happy to take anything that can be derefed into the correct types, but that's kind of a pain to model properly in the type system.

> A Store will remain always empty if its stream is never polled, I'm unsure how the subscription not producing items is more of a foot gun than that?

I'd generally expect (when dealing with async stuff) that a sync function operates on the current state, while an async function (or stream) would try to make progress. Implicitly waiting for another external thing to make said progress exposes the user to deadlocks which are a pain to debug.

We might be able to get around that by having the poller try to "steal" the backing stream, but I suspect that would cause a lot of thrashing.

> The lagging concern seems reasonable; I'm not sure of the best way to solve it. Calling stream.subscribable() could return a wrapper over the original stream and a single subscription stream in a tuple since the primary uses cases I see would be for a very small number of subscribers. The downside of that approach is each additional stream is another layer on top of the original stream.

I think that would still cause the same issue, you still have a "leader" stream and a "follower" stream, and you need to decide what happens when one consumer is making progress faster than the other.

The other way to handle this would be to pause the faster consumers to let the slower ones catch up. But this also means that we're back to worrying about even more deadlocks.

---

Maybe I'm overblowing the deadlock issue itself, and it'd be more pressing to make sure we have proper diagnostics for it when we suspect it. I don't know.

**Comment by nightkr** [comment]
Hm, as far as I can tell, `subscribe_ok()` is "just" propagated as `subscribe_all()`/`subscribe_namespaced()` in that PR, I'd be curious to understand how these are intended to be used in turn.

**Comment by danrspencer** [comment]
> Hm, as far as I can tell, `subscribe_ok()` is "just" propagated as `subscribe_all()`/`subscribe_namespaced()` in that PR, I'd be curious to understand how these are intended to be used in turn.

If you go back to the first commits on the PR I had an implementation which allowed a controller to run from the same stream of events as a store. I encountered the problem that the `applier` essentially ate the watcher stream, so you lost the original stream if you wanted to add logging or anything to it as it was consumed. I dropped that part of the PR for now because there was a lot going on in one PR and it seemed like it good be tackled in a few separate parts. 

The current `subscribe_all` and `subscribe_namespaced` are more brain storming that concrete suggestions on how a controller could be setup from a SharedStore to solve the split brain problem that can (hypothetically) currently occur.

It's entirely possible I'm going down the wrong path here and the correct solution is to stop the `applier` eating the `watcher` stream, or let it eat the stream and accept you can only have a single controller per stream and that you lose access to the watch stream when it's setup.

**Comment by danrspencer** [comment]
> I'd generally expect (when dealing with async stuff) that a sync function operates on the current state, while an async function (or stream) would try to make progress. Implicitly waiting for another external thing to make said progress exposes the user to deadlocks which are a pain to debug.

I think if it were an async function that would just deadlock I'd agree, but for a stream it feels entirely reasonable for it to just poll pending if there are no events on it yet. If Kubernetes has supplied no events then the `watcher` itself will not progress.

**Comment by nightkr** [comment]
> I think if it were an async function that would just deadlock I'd agree, but for a stream it feels entirely reasonable for it to just poll pending if there are no events on it yet. If Kubernetes has supplied no events then the watcher itself will not progress.

Rust streams are pulled, not pushed, just like futures.

> If you go back to the first commits on the PR I had an implementation which allowed a controller to run from the same stream of events as a store.

Ideally I think we'd share the same store too. I believe Go's controller library basically creates a new local stream by watching the store, something like that might be viable for us too.

**Comment by danrspencer** [comment]
> Ideally I think we'd share the same store too. I believe Go's controller library basically creates a new local stream by watching the store, something like that might be viable for us too.

The initial implementation I was playing with did have the Controller sharing the same store too. 

Wouldn't a new stream created by watching the store have exactly the same problem you're worried about with this PR? If you didn't consume the original watch stream then the stream watching the store would also never progress. Apologies if I'm completely missing some nuance in the details here 😅.

**Comment by nightkr** [comment]
You're right. I don't know, that'll probably just end up being a fact of life that we'll have to live with (and warn about in the documentation).

**Comment by danrspencer** [comment]
I've added in an update where the Stream wraps the events in Arcs which elevates some of the concerns discussed; some additional thoughts on it that I wanted to discuss before I do the work:

1. This makes the subscribe streams no longer compatible with `applier` without additional work. My suggestion would be to push the `Arc` further up so that the initial `watcher` function wraps all items in an `Arc` and then cascade that change down into all of the consumers.

2. I considered an implementation where the output stream of `StreamSubscribe` would be an other receiver. Every subscription and the `StreamSubscribe` stream would then check themselves when polled, if pending they'd poll the original stream to see if there's a new item that can be pulled into the channel. I decided against it because while it would fix the blocking issue, if there is a slow consumer then the `StreamSubscribe` itself may miss events; I think it makes more sense that the output stream of `StreamSubscribe` is considered the original and should always get everything, at the expense of forcing it to be consumed for subscribers to get their events.

File: kube-runtime/src/utils/stream_subscribe.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        assert_eq!(poll!(rx_s2.next()), Poll::Pending, "rx_s2 - pending");
+
+        for (index, item) in items.into_iter().enumerate() {
+            let expected = Poll::Ready(Some(Arc::new(item)));
+            assert_eq!(poll!(rx.next()), expected, "rx - {}", index);
+            assert_eq!(poll!(rx_s1.next()), expected, "rx_s1 - {}", index);
+            assert_eq!(poll!(rx_s2.next()), expected, "rx_s2 - {}", index);
+        }
+
+        // Ensure that if the stream is closed, all subscribers are closed
```
Comment (clux) [inline kube-runtime/src/utils/stream_subscribe.rs:109]: i'm curious to see what happens now if you test with a stream larger than the broadcast capacity (e.g. `(3..5).collect::<Vec<_>>()`), and try to exhaust rx before you start pulling from `rx_s1` what happens? 
presumably `this.sender.send(Some(item.clone())).ok();` above causes us to block from pulling `rx`?
worth testing this and documenting it if that's the case (with a big warning like; all subscribers must be polled)

File: kube-runtime/src/utils/watch_ext.rs [hunk 2]
```diff
@@ -36,5 +37,16 @@ pub trait WatchStreamExt: Stream {
     {
         EventFlatten::new(self, true)
     }
+
+    /// Create a [`StreamSubscribe`] from a [`watcher()`] stream
+    ///
+    /// Allows additional consumers to subscribe to the stream, without consuming the stream itself.
+    fn subscribable<K>(self) -> StreamSubscribe<Self>
```
Comment (clux) [inline kube-runtime/src/utils/watch_ext.rs:44]:
```suggestion
    fn subscribe<K>(self) -> StreamSubscribe<Self>
```

**Review by clux — COMMENTED** [review]
I think this looks great with the Arc, but want to document and test the edge cases here.

**Comment by clux** [comment]
> 1. This makes the subscribe streams no longer compatible with `applier` without additional work, which was the original motivation of this functionality. My suggestion would be to push the `Arc` further up so that the initial `watcher` function wraps all items in an `Arc` and then cascade that change down into all of the consumers.

That is possible.. but probably a pretty extensive change. I would want to see how much this affects the examples and general codebase before commit to an answer here. It is also possible that we make a new watcher like thing does this to let users move over in a non-breaking way somehow (we also have another generic extension in the works for watcher to allow doing watches on PartialObjectMeta types so there's more than one need) - but that does also limit the immediate controller usage (but we kind of also need some redesigning here to handle input streams so i think that is potentially OK).

**Comment by danrspencer** [comment]
> > 1. This makes the subscribe streams no longer compatible with `applier` without additional work, which was the original motivation of this functionality. My suggestion would be to push the `Arc` further up so that the initial `watcher` function wraps all items in an `Arc` and then cascade that change down into all of the consumers.
> 
> That is possible.. but probably a pretty extensive change. I would want to see how much this affects the examples and general codebase before commit to an answer here. It is also possible that we make a new watcher like thing does this to let users move over in a non-breaking way somehow (we also have another generic extension in the works for watcher to allow doing watches on PartialObjectMeta types so there's more than one need) - but that does also limit the immediate controller usage (but we kind of also need some redesigning here to handle input streams so i think that is potentially OK).

I guess an alternative would be to push the change down to the `applier` first. I imagine 99% of people are just using `Controller::new` so the wrapping of the events in an `Arc` could be wrapped inside that?

File: kube-runtime/src/utils/stream_subscribe.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+                Poll::Ready(Some(Ok(Arc::new(item)))),
+                "rx_s1"
+            );
+        }
+    }
+
+    #[tokio::test]
+    async fn if_the_subscribers_lag_they_get_a_lagged_error_as_the_next_event() {
+        // The broadcast channel rounds the capacity up to the next power of two.
+        let max_capacity = CHANNEL_CAPACITY.next_power_of_two();
```
Comment (clux) [inline kube-runtime/src/utils/stream_subscribe.rs:169]: afaikt https://doc.rust-lang.org/std/primitive.usize.html#method.next_power_of_two does not strictly increase the CHANNEL_CAPACITY, particularly if it already is a power of 2

Comment (clux) [reply kube-runtime/src/utils/stream_subscribe.rs:169]: ...ah, but you are overflowing by 5. nevermind!

File: kube-runtime/src/utils/stream_subscribe.rs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+/// An error returned from the inner stream of a [`StreamSubscribe`].
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum Error {
+    /// The subscriber lagged too far behind. Polling again will return
+    /// the oldest event still retained.
+    ///
+    /// Includes the number of skipped events.
+    Lagged(u64),
+}
```
Comment (clux) [inline kube-runtime/src/utils/stream_subscribe.rs:81]: what happens in the `Closed` case from https://docs.rs/tokio/latest/tokio/sync/broadcast/error/enum.RecvError.html here?
presumably it's the return `None` case to indicate a closed stream?

File: kube-runtime/src/utils/stream_subscribe.rs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        for _ in events {
+            rx.next().await;
+        }
+
+        assert_eq!(
+            poll!(rx_s1.next()),
+            Poll::Ready(Some(Err(Error::Lagged(overflow as u64)))),
+        );
+
+        let expected_next_event = overflow;
```
Comment (clux) [inline kube-runtime/src/utils/stream_subscribe.rs:190]: so we've gone al the way to the end to value 128+5 thus erasing the first 5.
good test, and nicely documented in the interface 👍

File: kube-runtime/src/utils/stream_subscribe.rs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        pin_mut!(rx_s1);
+        pin_mut!(rx_s2);
+
+        // Subscribers are pending until we start consuming the stream
+        assert_eq!(poll!(rx_s1.next()), Poll::Pending, "rx_s1");
+        assert_eq!(poll!(rx_s2.next()), Poll::Pending, "rx_s2");
+
+        for item in events {
+            assert_eq!(poll!(rx.next()), Poll::Ready(Some(Arc::new(item))), "rx");
+            let expected = Poll::Ready(Some(Ok(Arc::new(item))));
```
Comment (clux) [inline kube-runtime/src/utils/stream_subscribe.rs:131]: these lines almost completely cover the various cases of "subscribers is ahead of the main stream"
i think one extra assert in the for loop (try to poll `rx_s1` before `rx`) would completely close it.

File: kube-runtime/src/utils/watch_ext.rs [hunk 7]
```diff
@@ -36,5 +37,16 @@ pub trait WatchStreamExt: Stream {
     {
         EventFlatten::new(self, true)
     }
+
+    /// Create a [`StreamSubscribe`] from a [`watcher()`] stream
+    ///
+    /// Allows additional consumers to subscribe to the stream, without consuming the stream itself.
```
Comment (clux) [inline kube-runtime/src/utils/watch_ext.rs:43]: Same doc comment as on `StreamSubscribe` should be present here as this is likely the interface users go through.

Here it also needs an example.

**Review by clux — COMMENTED** [review]
tests are good now 👍 
minor nits on comments, will leave another comment on integration after.

File: kube-runtime/src/utils/stream_subscribe.rs [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+    /// Subscribe to events from this stream
+    #[must_use = "streams do nothing unless polled"]
+    pub fn subscribe(&self) -> impl Stream<Item = Result<Arc<S::Item>, Error>> {
+        stream::unfold(self.sender.subscribe(), |mut rx| async {
+            match rx.recv().await {
+                Ok(Some(obj)) => Some((Ok(obj), rx)),
+                Err(RecvError::Lagged(amt)) => Some((Err(Error::Lagged(amt)), rx)),
+                _ => None,
```
Comment (clux) [inline kube-runtime/src/utils/stream_subscribe.rs:45]: Oh, we need to handle errors better here i think;
- if the watcher emits an error here, it will restart, but we will just close the stream

so either we need to return a watcher::Error and union in a Lagged error into that enum, or we need to have our error here be a union of our new error and watcher::Error

Comment (clux) [reply kube-runtime/src/utils/stream_subscribe.rs:45]: ah, wait, `watcher` catches all of its errors though and retries.. maybe the `watcher` interface is a bit misleading here..

**Comment by clux** [comment]
> I guess an alternative would be to push the change down to the `applier` first. I imagine 99% of people are just using `Controller::new` so the wrapping of the events in an `Arc` could be wrapped inside that?

As it stands we are creating the distinction between two types of streams:

- a root stream of watcher events (not arcd, currently compatible with everything)
- an arcd subscriber stream of watcher events (arcd, not compatible with anything atm)

Currently, this distinction makes it hard for us to use (even within controller i think):

- `WatchStreamExt` will not be compatible; subscribers can't use `applied_objects()` style filters on a subscribed stream
- Subscribers have fundamentally different errors also - a LagError has to be handled whereas normally there are no errors bubbled up from a watcher stream

as a longer term solution; I'm wondering if this should be Arc'd at the start and we should start sketching out a more comprehensive/configurable `Watcher` struct that can help us solve other issues watcher at the same time;

- Arc's the items
- unifies the error enum (possibly unioning in Subscriber::Error)
- allows configuring the watcher errors - currently i don't believe we bubble up a single error :|
- pagination on initial list call to handle busier resources better
- abstracting over `watch` and `watch_metadata` to allow metadata watchers

I'll write up some thoughts about that in a separate issue.

In either case, I think as it stands this PR is OK without further integration for now provided it's feature flagged while we sketch these things out; maybe under feat=`unstable_runtime_subscribe` inside a `unstable_runtime` master feat.

File: kube-runtime/src/utils/watch_ext.rs [hunk 9]
```diff
@@ -4,6 +4,7 @@ use crate::{
 };
 use backoff::backoff::Backoff;
 
+use crate::utils::stream_subscribe::StreamSubscribe;
```
Comment (clux) [inline kube-runtime/src/utils/watch_ext.rs:7]: This needs a feature gate.

I would add unstable features to the documentation build as well. Those are listed in `[package.metadata.docs.rs]` in kube/Cargo.toml. To simplify this it might be easiest if there was one super-feature called unstable_runtime that we could add to that.

Comment (danrspencer) [reply kube-runtime/src/utils/watch_ext.rs:7]: I added in. 

```toml
[features]
unstable_runtime = ["unstable_runtime_subscribe"]
unstable_runtime_subscribe = []
```

Is this the "super-feature" you envisaged, or would that live in the the `kube/cargo.toml`?

Comment (clux) [reply kube-runtime/src/utils/watch_ext.rs:7]: e.g.

```toml
unstable-runtime = ["kube-runtime/unstable-runtime"]
```

File: kube-runtime/src/utils/mod.rs [hunk 10]
```diff
@@ -3,11 +3,14 @@
 mod backoff_reset_timer;
 mod event_flatten;
 mod stream_backoff;
+mod stream_subscribe;
```
Comment (clux) [inline kube-runtime/src/utils/mod.rs:6]:
```suggestion
#[cfg(feature = "unstable_runtime_subscribe")]
mod stream_subscribe;
```

should fix the last clippy issues

File: kube-runtime/src/utils/stream_subscribe.rs [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+impl std::error::Error for Error {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::{pin_mut, poll, stream, StreamExt};
+
+    #[tokio::test]
+    async fn stream_subscribe_continues_to_propagate_values() {
```
Comment (clux) [inline kube-runtime/src/utils/stream_subscribe.rs:104]: hm, how are you running these tests?
i was on your branch trying `cargo test -p kube-runtime --features=unstable-runtime-subscribe --lib` and i'm not seeing any `subscribe` results, not even with `--all-features` makes me think that some of the feature selection stuff is not working correctly

they don't show up in CI unit tests either: https://github.com/kube-rs/kube/actions/runs/4205152176/jobs/7296822788 but that's probably just a missing `--all-features` to add to https://github.com/kube-rs/kube/blob/main/.github/workflows/ci.yml#L46-L47 once we figure out why they don't show up at all)

**Review by clux — APPROVED** [review]
Awesome work, thank you so much. This is all good to me, apart from the **DCO** check that needs a `--signoff`, but beyond that I am all happy to merge.

This PR has also helped me clarify some thoughts on how we can do unstable features (yes, people can accidentally enable these, but our most advanced use generally makes the most sense as a direct dependency so people generally have to opt-in to this).
